src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -0,0 +1,355 @@
+
+        // IDisposable
+
+        public void Dispose()
+        {
+        }
+
+        private static class EmptyTensor
+        {
+#pragma warning disable CA1825 // this is the implementation of Tensor.Empty<T>()
+            internal static readonly Tensor<T> Value = new Tensor<T>();
+#pragma warning restore CA1825
+        }
+        // ITensor
+
+        public static Tensor<T> Empty => EmptyTensor.Value;
+
+        public bool IsEmpty { get { return _lengths.Length == 0; } }
+
+        public bool IsPinned { get { return _isPinned; } }
+
+        public int Rank { get { return _lengths.Length; } }
+
+        /// <summary>
+        /// The number of items in the span.
+        /// </summary>
+        internal nint LinearLength
+        {
+            get => _linearLength;
+        }
+        // REIVEW: SHOULD WE CALL THIS DIMS OR DIMENSIONS?
+        /// <summary>
+        /// Gets the length of each dimension in this <see cref="Tensor{T}"/>.
+        /// </summary>
+        public ReadOnlySpan<nint> Lengths { get { return _lengths; } }
+        public ReadOnlySpan<nint> Strides { get { return _strides; } }
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[params nint[] indices] => ref this[indices.AsSpan()];
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[ReadOnlySpan<nint> indices] => ref AsSpan()[indices];
+
+        // REVIEW: NOT IN API DOC BUT IN NOTEBOOK AND IN OTHER FRAMEWORKS
+        // REVIEW: WE WILL WANT THIS CHANGED FROM A BOOL TO SOME FILTER EXPRESSION.
+        public Tensor<T> this[Tensor<bool> filter]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (filter.Lengths.Length != Lengths.Length)
+                    throw new ArgumentOutOfRangeException(nameof(filter), "Number of dimensions to slice does not equal the number of dimensions in the span");
+
+                for (var i = 0; i < filter.Lengths.Length; i++)
+                {
+                    if (filter.Lengths[i] != Lengths[i])
+                        ThrowHelper.ThrowArgumentOutOfRangeException();
+                }
+
+                var srcSpan = MemoryMarshal.CreateSpan(ref _values[0], (int)_linearLength);
+                var filterSpan = MemoryMarshal.CreateSpan(ref filter._values[0], (int)_linearLength);
```

Comment:
In general, please only use `var` when the type is actually written on the right-hand side.